include/swift/Runtime/Concurrent.h

```diff
@@ -406,6 +409,102 @@ class ConcurrentMap
   }
 };
 
+
+/// An append-only array that can be read without taking locks. Writes
+/// are still locked and serialized, but only with respect to other
+/// writes.
+template <class ElemTy> struct ConcurrentReadableArray {
+private:
+  /// The struct used for the array's storage. The `Elem` member is
+  /// considered to be the first element of a variable-length array,
+  /// whose size is determined by the allocation. The `Capacity` member
+  /// from `ConcurrentReadableArray` indicates how large it can be.
+  struct Storage {
+    std::atomic<size_t> Count;
+    typename std::aligned_storage<sizeof(ElemTy), alignof(ElemTy)>::type Elem;
+    
+    ElemTy *data() {
+      return reinterpret_cast<ElemTy *>(&Elem);
+    }
+  };
+  
+  size_t Capacity;
+  std::atomic<size_t> ReaderCount;
+  std::atomic<Storage *> Elements;
+  Mutex WriterLock;
+  std::vector<Storage *> FreeList;
+  
+  Storage *allocate(size_t capacity) {
+    auto size = sizeof(Storage) + (capacity - 1) * sizeof(Storage().Elem);
+    auto *ptr = reinterpret_cast<Storage *>(malloc(size));
+    if (!ptr) swift::crash("Could not allocate memory.");
+    ptr->Count.store(0, std::memory_order_relaxed);
+    return ptr;
+  }
+  
+  void deallocate(Storage *storage) {
+    if (storage == nullptr) return;
+    
+    auto *data = storage->data();
+    for (size_t i = 0; i < storage->Count; i++) {
+      data[i].~ElemTy();
+    }
+    free(storage);
+  }
+  
+public:
+  void push_back(const ElemTy &elem) {
+    ScopedLock guard(WriterLock);
+    
+    auto *storage = Elements.load(std::memory_order_relaxed);
+    auto count = storage ? storage->Count.load(std::memory_order_relaxed) : 0;
+    if (count >= Capacity) {
+      auto newCapacity = std::max((size_t)16, count * 2);
+      auto *newStorage = allocate(newCapacity);
+      if (storage) {
+        std::copy(storage->data(), storage->data() + count, newStorage->data());
+        FreeList.push_back(storage);
+      }
+      
+      storage = newStorage;
+      Capacity = newCapacity;
+      Elements.store(storage, std::memory_order_release);
+    }
+    
+    auto Count = storage->Count.load(std::memory_order_relaxed);
```

Comment:
Does this need to be reloaded? I think the other `count` will have the same value purpose, because there can't be concurrent modifications to `storage->Count`.